proof/refine/ARM/ArchInvsLemmas_H.thy

```diff
@@ -481,40 +420,55 @@ lemma objBitsT_koTypeOf:
+  p && ~~ mask (vs_ptr_align ko) = p' && ~~ mask (vs_ptr_align ko)\<rbrakk>
+  \<Longrightarrow> m p' = Some ko "
+  apply (clarsimp simp:vs_valid_duplicates'_def)
+  apply (drule_tac x = p in spec)
+  apply auto
+  done
 
-qualify AARCH64_H (in Arch)
+lemma page_directory_pde_atI':
+  "\<lbrakk> page_directory_at' p s; x < 2 ^ pageBits \<rbrakk> \<Longrightarrow> pde_at' (p + (x << 2)) s"
+  by (simp add: page_directory_at'_def pageBits_def)
 
-(*
-  Then idea with this class is to be able to generically constrain
-  predicates over pspace_storable values to are not of type VCPU,
-  this is useful for invariants such as obj_at' that are trivially
-  true (sort of) if the predicate and the function (in the Hoare triple)
-  manipulate different types of objects
-*)
-
-class no_vcpu = pspace_storable +
-  assumes not_vcpu: "koType TYPE('a) \<noteq> ArchT AARCH64_H.VCPUT"
-
-instance tcb              :: no_vcpu by intro_classes auto
-instance endpoint         :: no_vcpu by intro_classes auto
-instance notification     :: no_vcpu by intro_classes auto
-instance cte              :: no_vcpu by intro_classes auto
-instance user_data        :: no_vcpu by intro_classes auto
-instance user_data_device :: no_vcpu by intro_classes auto
-
-end_qualify
-
-instantiation AARCH64_H.asidpool :: no_vcpu
-begin
-interpretation Arch .
-instance by intro_classes auto
-end
+lemma page_table_pte_atI':
+  "\<lbrakk> page_table_at' p s; x < 2^(ptBits - 2) \<rbrakk> \<Longrightarrow> pte_at' (p + (x << 2)) s"
+  by (simp add: page_table_at'_def pageBits_def ptBits_def pteBits_def)
+
+lemma pdBits_eq: "pdBits = pd_bits"
+  by (simp add: pd_bits_def pdBits_def pdeBits_def pageBits_def)
+
+lemma ptBits_eq: "ptBits = pt_bits"
+  by (simp add: pt_bits_def ptBits_def pteBits_def pageBits_def)
+
+lemma largePagePTE_offset_eq:  "largePagePTE_offsets = largePagePTEOffsets"
+  by (simp add: largePagePTE_offsets_def largePagePTEOffsets_def pteBits_def)
+
+lemma superSectionPDE_offsets_eq:  "superSectionPDE_offsets = superSectionPDEOffsets"
+  by (simp add: superSectionPDE_offsets_def superSectionPDEOffsets_def pdeBits_def)
+
+lemma valid_arch_state'_valid_pde_mappings'[elim!]:
+  "valid_arch_state' s \<Longrightarrow> valid_pde_mappings' s"
+  by (simp add: valid_arch_state'_def)
+
+lemma valid_pde_mapping'_simps[simp]:
+  "valid_pde_mapping' offset (InvalidPDE) = True"
+  "valid_pde_mapping' offset (SectionPDE ptr a b c d e w)
+   = valid_pde_mapping_offset' offset"
+  "valid_pde_mapping' offset (SuperSectionPDE ptr a' b' c' d' w)
+   = valid_pde_mapping_offset' offset"
+  "valid_pde_mapping' offset (PageTablePDE ptr x z'')
+   = valid_pde_mapping_offset' offset"
+  by (clarsimp simp: valid_pde_mapping'_def)+
+
+(* FIXME arch-split: possibly can go into Invariants_H *)
+lemma valid_pspaceI'[intro]:
+  "\<lbrakk>valid_objs' s; pspace_aligned' s; pspace_distinct' s; valid_mdb' s; no_0_obj' s\<rbrakk>
+   \<Longrightarrow> valid_pspace' s"
+  unfolding valid_pspace'_def by simp
```

Comment:
I think that might work, yes. Would at least be nice. Ok to leave for the next pass, though.

Reply:
Will leave for next pass because the `[intro]` might trigger needing fixups on other arches.